### ISSUE TITLE
Expose a closure on the task to manipulate the SourceReport before publishing

### ISSUE THREAD

**Issue by nhoughto** [nbaztec/coveralls-jacoco-gradle-plugin#27]
I am trying to set things up in a monorepo with lots of different projects and trying to create a particular result in coveralls wrt file paths, currently source files are emitted with paths relative to the project itself with no way to change them before they are published. It would be great if the task could be configured with a closure that takes maybe the `List<SourceReport>` and have the closure execute before the publish if its configured, this would give ultimate flexibility to the caller with little effort.

**Comment by nhoughto** [comment]
Be nice to have a closure to manipulate the `Request` object before publishing too, just had a problem where the `jobId` needed to be enriched with another key (because doing many reports in one CI job).

**Comment by nhoughto** [comment]
Interest in a PR for this? happy to contribute

**Comment by nbaztec** [comment]
Could you please elaborate on the use case for the feature request.

For example: Under what circumstances would one require to change the file paths and/or the job id for a single run?

**Comment by nhoughto** [comment]
So in a mono repo with many projects the reported sources and their paths as reported to coveralls is relative to the project so, “src/main/java/Blah.java”, but I want the path to include the full repository path of “project/module/name/src/main/java..” so that in coveralls I can see all the projects and their sources in the correct location, not just a mix of all the source files without their project names.
Also why not? Like it just makes it more flexible that makes it more useful to more people 🤷🏻‍♂️

**Comment by nbaztec** [comment]
Ideally the source path should *just work* when run from a Gradle task and the path should be picked up correctly in coveralls (incorrect path on coveralls will lead to coverage not being reported correctly). 

Perhaps it's a bug that you encountered that needs fixing. I'd suggest fixing it in code than providing a hook for users to fix themselves. Would be great if there's some sample repo where it's not working as intended so we can take a look together.

Thanks.

**Comment by nhoughto** [comment]
No bug, I’m just doing something most aren’t doing, trying to design the output of coveralls for a mono repo with hundreds of projects. The configured of the task having ultimate control as a possible escape hatch is useful when you are slightly off the beaten path

**Comment by nbaztec** [comment]
A mono repo could also work (provided all projects are using jacoco) - the jacoco reports can be merged together on the root level with a  root-level `build.gradle` into a single file as shown [here](https://github.com/nbaztec/coveralls-jacoco-gradle-plugin#kotlin-dsl-buildgradlekts)

Alternatively, since the plugin consumes the `xml` report - for a certain purpose the multiple `xml` report files can actually be merged manually via external script while prepending the project directory `proj1/src/kotlin` to the `package.name` xml elements. The format is like:
```xml
<package name="foo/bar">
  <sourcefile name="HelloKotlin.kt">
    ...
  </sourcefile>
</package>
```
and then setting the `reportPath` to this merged config and `reportSourceSets` to just the root of the mono.
The theory being if transformation is required, it can be done at the report/input level itself, before anything reaches the plugin.

Providing a hook seems quite hacky IMO and wouldn't be in line with the intended purpose of the plugin, since the data can be completely transformed effectively in the hook rendering all options/parsing redundant. The plugin in itself is simply reading *a* jacoco report file and reporting it to coveralls. In line with that, there are multiple examples in the README on how to merge reports.

I'll could perhaps try to assist you in finding a solution but unfortunately wouldn't feel comfortable adding a free-for-all hook.

**Comment by nhoughto** [comment]
Hmm not all projects are jacoco, some are TypeScript and C#, but I could probably use a single way for all jacoco projects. The problem I found with the suggested 'combine and report once in root' was that you then couldn't use different project sourceSets effectively. If each project creates a jacoco report and then a single task at the root aggregates, enriches the package name path with extra path info then runs the reporting plugin, the plugin will check if all the files exist in the repo (which is fine) but it falls down when you have multiple sourceSets. At a project level if i'm trying to find file `src/main/java/Blah.java` that could be in the `main` sourceSet or the `generated` sourceSet, and the plugin would find that file because the paths are correct and it can search both sourceSets. If I aggregate things and report once from the root, the plugin will try and find `project/module1/src/main/java/Blah.java` which it won't find, as its a generated sourceSet under `project/module1/build/generated/sources/src/main/java/Blah.java` which it has no ability to search effectively anymore as i've told it the file is in `project/module1/src/main/java/Blah.java`. The detail around generation from a user POV is irrelevant since the IDE abstracts it away for the most part, and I definitely don't want to see it in Coveralls list of files as that would be messy and pushing implementation details into a coverage tool.. 

So in Gradle land there are many different ways to set things up and most of the time they aren't 'wrong' just different, and the ways things can be different increases with every Gradle version it feels like, so from a plugin POV to avoid having to explicitly support or not support those different patterns you support what you do today, but for everyone else you can use the closure as an escape hatch if you really need to. Majority of simple project users won't use it, but the advanced users will be thankful for it.

**Comment by nbaztec** [comment]
I see your point, however one thing to keep in mind is that coveralls will never report on generated files if they don't belong to the checked-in repository. 
Even in its current state, if jacoco reports coverage from any generated sources, the plugin would ignore them since no physical file exists within the actual git source (which is incidentally the `src` directory) - which is also what Coveralls reports on. The generated files are simply ignored. The merging approach would then still work.

Alternatively, perhaps if we add a *dry run* option and dump the request to a file rather than reporting to coveralls, perhaps, you can do the necessary transformation "off-site" and submit the merged payload to coveralls via a simple [http POST](https://github.com/nbaztec/coveralls-jacoco-gradle-plugin/blob/main/src/main/kotlin/CoverallsReporter.kt#L75) once done. Would that be an option?

I apologize, if I'm seeming a bit hesitant to simply add a hook which is trivial from a code point of view, but I'm merely trying to cover all aspects of the UX as it will be perceived for future users as well, where removing immutability can have confusing side-effects. I agree the Gradle ecosystem has this plethora of ways to do the same thing, which personally I believe is often a very confusing and bad API design, albeit the ecosystem has adapted accordingly to it. 

And I appreciate the use case & examples that you are providing in helping me understand the requirement.

**Comment by nhoughto** [comment]
Yep you could do a convert task and a publish task, so if you want to manipulate and self-publish that would be your mechanism. Seems a bit more roundabout and more effort for users of the escape hatch than a simple closure that you can opt-in to but it would work. Could there be a convert, publish and a publish and convert task? So that I can re-use the publish logic?

So new task would be just a convert step (writes JSON to buildDir somewhere) then a second task to do the coveralls publish from the JSON. Then the existing task would wire both tasks together to create the existing functionality. Again seems more complex than a hook but maybe it suits purposes more 🤔

**Comment by nbaztec** [comment]
Oh that's weird of coveralls, in my tests everything in the build directory stayed out since it wasn't checked in the repo (perhaps they stopped synchronizing with the repo now). Generated files usually never contribute to the coverage for the actual source files in most environments.

If you put it like that then yes, it more convoluted. But if the plugin merely dumps the request to `build/coverallsRequest.json` then the user can have a custom gradle task to read all these files, merge them, then send it to coveralls with correct parameters. The transformation step is basically abstracted out to a custom task in itself. If the task executes per project then the merging is optional and the same file can be overwritten.

example:

```groovy
task mergeAndPostCoverage {
    doLast {
        def allFiles = new File(".").listFiles({it.name == 'coverallsRequest.json' } as FileFilter)
        def mergedFile = new File(project.buildDir, "mergedCoverallsRequest.json")

        // do something and write merged file

        OkHttpBuilder.configure {
            request.uri = 'https://coveralls.io'
        }.post {
            request.uri.path = '/api/v1/jobs'
            request.contentType = 'multipart/form-data'
            request.body = multipart {
                field 'name', 'json_file'
                part 'file', 'json_file', 'application/json', mergedFile
            }
            request.encoder 'multipart/form-data', OkHttpEncoders.&multipart
        }
    }
}
```

**Comment by nhoughto** [comment]
Yep a custom task could do it, I was making it sound complicated so you just did the hook =)
Dumping to JSON would allow me to solve my problem, so its better than nothing 👍

**Comment by nbaztec** [comment]
😄  Thanks for being understanding. 
I will keep an eye out if someone else also comes asking for a hook feature, then I'll definitely give you a shout out again :)

**Comment by nbaztec** [comment]
I've released `v1.2.6` with `dryRun` and `coverallsRequest` options. Let me know if it helped 👍

**Comment by nhoughto** [comment]
thanks will test

**Comment by nhoughto** [comment]
doing the publish myself was a little janky, had to add an http client to my script to do it and quite verbose, functional but ugly, have simple PR to make it possible to do via the plugin.
https://github.com/nbaztec/coveralls-jacoco-gradle-plugin/pull/32

**Comment by nbaztec** [comment]
`v1.2.7` released with changes from #32
